src/descendPeak.cpp

```diff
@@ -0,0 +1,143 @@
+// Pascal Maas - p.maas@lacdr.leidenuniv.nl
+
+#include <Rcpp.h>
+using namespace Rcpp;
+using namespace std;
+
+double weightedMean(NumericVector mzs, NumericVector ints) {
+    // Get the total intensity
+    double intensity = sum(ints);
+
+    // Calculate the weighted mass
+    int n = ints.length();
+    double total = 0;
+    for (int i = 0; i < n; i++) {
+        total += mzs[i] * ints[i] / intensity;
+    }
+    return total;
+}
+
+double descend(int centroid, NumericVector intensities,
+               NumericVector masses, double signalRatio) {
+
+    // Maximum deviation from the centroid index
+    int maxK = 30;
+
+    // Define a min and max index for the region where the region is not negative
+    // and does not exceed the size of the intensities vector, since we only
+    // have to consider centroid - maxK until centroid + maxK
+
+    // Ensures the index is at least 0
+    int minId = std::max(centroid - maxK, 0);
+
+    // Ensures the index is at most the size of the intensities vector - 1
+    int maxId = std::min(centroid + maxK, (int) intensities.size() - 1);
+
+    // Get indices of this region
+    Rcpp::Range idx = Rcpp::Range(minId, maxId);
+
+    // Subset intensities and masses with the defined region
+    NumericVector ints = intensities[idx];
+    NumericVector mass = masses[idx];
+
+
+    // Find the new location of the centroid, as the centroid can be shifted by
+    // subsetting the region. By substracting the minimum value, the original
+    // centroid will now refer to the new centroid location.
+    centroid = centroid - minId;
+    double centroidValue = ints[centroid];
+
+    // Calculate the allowed limits within the region
+    // Update the left and right limits with new idx, as long as the values are
+    // within the thresholds of percentage and are not rising.
+    int leftLimit = centroid;
+
+    // Start looping from centroid to the left (negative)
+    for (int i = centroid - 1; i >= 0; i--) {
+        // Get current and previous values
+        double previous = ints[i + 1];
+        double current = ints[i];
+
+        // Good signal should be above the centroid / signal ratio
+        bool goodSignal =  current / centroidValue > signalRatio;
+
+        // Compare current to previous value to check for a rising signal
+        bool risingSignal = current >= previous;
```

Comment:
Would this be a good test? Already a small plateau (2 indices of the same intensity value) would break here, even if it goes down after the overnext index.

Reply:
I do believe that is the default behaviour in the original MSnbase version here: https://github.com/lgatto/MSnbase/blob/1338323345d2a6cf856f1819dfb365300ab6d9d5/R/functions-Spectrum.R#L781-L788

I'm not sure about the current version in MsCoreUtils, but I could implement the `stopAtTwo` behaviour or perhaps stop at any given `k` if desired?